flow/display_list_utils.h

```diff
@@ -314,80 +343,233 @@ class DisplayListBoundsCalculator final
+  // An intermediate implementation class that handles keeping
+  // a local accumulator for the layer, used by both |RootLayerData|
+  // and |SaveLayerData|.
+  class AccumulatorLayerData : public LayerData {
    public:
-    SaveLayerInfo(BoundsAccumulator* accumulator, const SkMatrix& matrix);
-    virtual ~SaveLayerInfo() = default;
+    BoundsAccumulator* accumulatorForLayer() override {
+      return &layer_accumulator_;
+    }
 
-    BoundsAccumulator* save() override;
-    BoundsAccumulator* restore() override;
+    SkRect getLayerBounds() override {
+      // Even though this layer might have been flooded, we still
+      // accumulate what bounds we have as the flooded condition
+      // may be a nop at a higher level and we at least want to
+      // account for the bounds that we do have.
+      return layer_accumulator_.getBounds();
+    }
 
    protected:
-    BoundsAccumulator layer_accumulator_;
-    const SkMatrix matrix_;
+    using LayerData::LayerData;
+    ~AccumulatorLayerData() = default;
 
    private:
-    FML_DISALLOW_COPY_AND_ASSIGN(SaveLayerInfo);
+    BoundsAccumulator layer_accumulator_;
   };
 
-  class SaveLayerWithPaintInfo : public SaveLayerInfo {
+  // Used as the initial default layer info for the Calculator.
+  class RootLayerData : public AccumulatorLayerData {
    public:
-    SaveLayerWithPaintInfo(DisplayListBoundsCalculator* calculator,
-                           BoundsAccumulator* accumulator,
-                           const SkMatrix& save_matrix,
-                           const SkRect* bounds,
-                           const SkPaint& save_paint);
-    virtual ~SaveLayerWithPaintInfo() = default;
+    RootLayerData() : AccumulatorLayerData(nullptr) {}
+    ~RootLayerData() = default;
 
-    BoundsAccumulator* restore() override;
+   private:
+    FML_DISALLOW_COPY_AND_ASSIGN(RootLayerData);
+  };
 
-   protected:
-    DisplayListBoundsCalculator* calculator_;
+  // Used for |save|
+  class SaveData : public LayerData {
+   public:
+    using LayerData::LayerData;
+    ~SaveData() = default;
+
+   private:
+    FML_DISALLOW_COPY_AND_ASSIGN(SaveData);
+  };
 
-    std::optional<SkRect> bounds_;
-    SkPaint paint_;
+  // Used for |saveLayer|
+  class SaveLayerData : public AccumulatorLayerData {
+   public:
+    SaveLayerData(BoundsAccumulator* outer,
+                  sk_sp<SkImageFilter> filter,
+                  bool paint_nops_on_transparency)
+        : AccumulatorLayerData(outer), layer_filter_(std::move(filter)) {
+      if (!paint_nops_on_transparency) {
+        set_flooded();
+      }
+    }
+    ~SaveLayerData() = default;
+
+    SkRect getLayerBounds() override {
+      SkRect bounds = AccumulatorLayerData::getLayerBounds();
+      if (!getFilteredBounds(bounds, layer_filter_.get())) {
+        set_flooded();
+      }
+      return bounds;
+    }
 
    private:
-    static constexpr SkRect kMissingBounds = SkRect::MakeWH(-1, -1);
+    sk_sp<SkImageFilter> layer_filter_;
```

Comment:
Are we 100% sure that this will always get collected on the raster task runner?

Reply:
(This one is definitely not safe to collect on non-raster task runners)

Reply:
Since when? lib/ui/painting/image_filter.h holds them in an sk_sp and the SkPicture absolutely positively used to hold them in an sk_sp...?

Reply:
Are we just lucking out then? An Image filter will almost always have an SkImage inside, and that isn't safe to destroy on the UI task runner, which is why we use a SkiaGPUObject. There are race conditions involved though, so maybe it just works out most of the time?

/cc @jason-simmons or @chinmaygarde who might know something I'm missing here.

Reply:
yes - it looks like `flutter::ImageFilter` and `flutter::ImageFilterLayer` should be wrapping the `SkImageFilter` in a `SkiaGPUObject`.

`SkImageFilter` implementations (specifically `SkImageImageFilter`) may hold references to `SkImage` objects.  So these objects should be dereferenced on the raster thread.

Reply:
I'd like to hear more about this because I think we have a long history of storing these as just regular shared pointers.

On a related front, it looks like we store images as `flutter::SkiaGPUObject<SkImage>` in the lib/ui/painting/image.cc class, but when we add them to a display list or an SkPicture we extract the Skia object and just hold onto that instead. That's probably where there may be a bug lurking if anywhere...?

Reply:
Flutter doesn't expose any ImageFilter that holds a reference to an image. We have a native method to do so in the engine, but I don't see any calls to it from the framework...?

Reply:
Ah, the Dart Picture native wrapper holds the SkPicture in a `SkiaGPUObject`, but the SkPicture it holds can outlive that native object, so that seems to be an attempt to solve this problem, but I don't see how it is robust.

We currently only hold the DisplayList  in an sk_sp. We can upgrade that to a SkiaGPUObject. I'll file a bug on it.

Reply:
https://github.com/flutter/flutter/issues/90237

Reply:
How does our use of `SkPicture` work where you can embed an `SkImage` using an `sk_sp`, and elsewhere the `SkiaGPUObject` from whence it came could be reclaimed, but the `sk_sp` version is still referenced inside the `SkPicture`?

Reply:
Where do `ImageImageFilter`s come from anyway? The framework doesn't seem to have a method to make them.

Reply:
Interesting - the `flutter::ImageFilter` native peer class defines a method for constructing an `SkImageImageFilter`

But it looks like the engine Dart APIs do not actually expose that method to Dart code.

If `ImageFilter::initImage` is unused then it should be removed.  And if none of the `SkImageFilter` subclasses that the engine creates are holding on to any GPU resources, then the lack of the `SkiaGPUObject` wrapper is not an issue.

`SkiaGPUObject` is needed for objects that will be deleted on the UI thread but are referencing other objects that must be deleted on the raster thread.  It's used for Dart native peer objects that hold references to Skia objects whose destructors may call graphics APIs.

If an `sk_sp<SkPicture>` is guaranteed to be deleted on the raster thread then it does not need the `SkiaGPUObject` wrapper.

Reply:
`sk_sp<SkPicture>` is not guaranteed to be deleted on the raster thread, it is a regular shared pointer.

`SkiaGPUObject<SkPicture>` is guaranteed to be deleted on the raster thread. `flutter::Picture` holds a `SkiaGPUObject<SkPicture>`. `flutter::PictureLayer` also holds a copy via `SkiaGPUObject`. If either of them hand out an extra copy of the skia version of the picture, then we need to hope that that reference gets unrefed before either of those objects are unrefed.

In either case, it's not clear that `SkPicture` itself needs to be deleted on the raster thread unless it holds a pointer to a Skia object that also needs to be deleted on the raster thread (SkImage or anything that holds an SkImage - anything else?).

But what about an `SkImage` that is held by an `SkPicture`? If the `flutter::Image` goes away before the `flutter::Picture` then its `SkiaGPUObject` will deref and potentially queue the `SkImage` for deletion. When that `SkiaGPUObject` reaches the raster thread, it will get deref'd, but it won't get freed! Instead, the skia object itself will be held by the `SkPicture`. Fortunately, if we also delete the SkPicture on the raster thread, then it should work out.

For now, we could add a `SkiaGPUObject` wrapper for DisplayList objects in `flutter::Picture` and `flutter::PictureLayer`. I submitted an Issue for that (mentioned above: https://github.com/flutter/flutter/issues/90237) and will take care of that separately.

Reply:
Is there a reason that the GPU resources in the SkImage object can't be taught to manage this themselves? This seems like a house of cards built around an object that might potentially hold thread-constrained resources causing anyone who might touch one to have to jump through hoops. The GPU resource itself should protect itself without any outside burden...

Reply:
The `SkiaGPUObject<SkPicture>` itself will be deleted on the UI thread.  `SkiaGPUObject<SkPicture>` wraps an `sk_sp<SkPicture>`, and `SkiaGPUObject's` destructor will queue deletion of the `sk_sp<SkPicture>` to the raster thread.

For a specific `sk_sp<SkPicture>` instance it may be the case that it is used exclusively by code running on the raster thread.  If so, then that `sk_sp<SkPicture>` will be destructed safely on the raster thread and `SkiaGPUObject` is therefore unnecessary.

If there is a chance that an `sk_sp<SkPIcture>` instance might be destructed on the UI thread, then it must be wrapped with a `SkiaGPUObject` so that any resources such as `SkImages` referenced by the `SkPicture` will be deleted on the raster thread.

The same would apply to a `DisplayList` if it could be destructed on the UI thread.

Reply:
Skia considers it the application's responsibility to configure threads to access any required graphics APIs.

On a platform such as OpenGL/EGL, the raster thread has a context for calling OpenGL APIs but the UI thread does not.  So the engine needs to ensure that any code that might invoke those APIs (such as Skia object destructors) runs on the raster thread.

Reply:
It looks like SkImageFilter is an SkFlattenable, so I guess it's ok? A quick look at the headers related to it don't seem to show it retaining any GPU resources.

Reply:
And @flar - we have `PictureLayer`s create their own SkiaGPUObject wrapping the SkPicture for that reason.

Reply:
For the short term I've fixed this under https://github.com/flutter/engine/pull/28679

Eventually we will overhaul how we deal with Skia objects inside the DisplayList and we may be able to relax this restriction on holding DisplayList objects in GPU object wrappers. For now, that PR should make DisplayList as robust as SkPicture is currently...

Reply:
Just to follow up here - after a discussion with Skia it looks like the intention is for all types to be thread-safe with respect to construction and deletion and they all but promise that for any type that can be held in an SkPaint, except that it is possible to run afoul of the one area of the API that can't (yet) make that promise - SkImage objects. They also largely all but make that promise for SkPicture except that it can also hold references to SkImage objects.

Fixing the SkImage issue was a planned goal for the next major revision, but they acknowledge that it is creating technical debt for some of their customers, including Flutter, so they are looking for opportunities for moving that time frame up. No promises on this front, but I wanted to document here what we know about the situation since there was speculation in this comment thread.